colossus/src/main/scala/colossus/protocols/http/HttpResponseParser.scala

```diff
@@ -3,68 +3,231 @@ package protocols.http
+    builder.putBytes(data.takeAll)
+    builder.append(HttpParse.NEWLINE)
+    DataBuffer(builder.result)
+  }
+}
+
+/** A Pipe that will take raw DataBuffers and add a http chunk header.  This is
+ * needed if you are streaming out a response that is not also being streamed
+ * in (like in a proxy)
+ *
+ * Right now this ends up copying the DataBuffer.
+ * 
+ */
+class ChunkEncodingPipe extends InfinitePipe[DataBuffer] {
+
+  override def push(data: DataBuffer) = whenPushable {
+    super.push(HttpChunk.wrap(data))
+  }
+
+  override def complete() {
+    super.push(DataBuffer(ByteString("0\r\n\r\n"))) match {
+      case PushResult.Full(trig) => trig.fill(complete)
+      case _ => super.complete()
+    }
+  }
+
+}
+
+
+/**
+ * A Pass-through pipe that parses chunk headers in the data stream and closes
+ * the pipe when the stream ends.  This pipe does not remove any header/footer
+ * information, making it ideal for use in a proxy
+ */
+class ChunkPassThroughPipe extends InfinitePipe[DataBuffer] {
+
+  //this parser is used to follow along with the chunks so we know when to
+  //close the stream, but we don't actually parse out the data since we're just
+  //passing everything through
+  private val chunkParser = intUntil('\r', 16) <~ byte |>{
+    case 0   => const(0) //notice there's no second \r\n, this is according to spec
+    case len => skip(len.toInt) <~ bytes(2) >> {_ => len.toInt}
   }
+
+  private val trailerParser = bytes(2)
+
+  private var parsingTrailer = false
+
+  override def push(data: DataBuffer): PushResult = whenPushable {
+    val (done, bytesRead) = data.peek{d => 
+      try {
+        if (!parsingTrailer) {
+          var chunkSize = Int.MaxValue
+          while(d.hasUnreadData && chunkSize != 0) {
+            chunkParser.parse(d).foreach{len =>
+              chunkSize = len
+            }
+          }
+          parsingTrailer = if (chunkSize == 0) true else false
+        }
+        if (parsingTrailer) {
+          if (trailerParser.parse(d).isDefined) {
+            true
+          } else {
+            false
+          }
+        } else {
+          false
+        }        
+      } catch {
+        case p: ParseException => {
+          terminate(p)
+          false //this will result in a push failure
+        }
+      }
+    }
+    if (done) {
+      super.push(DataBuffer(data.take(bytesRead)))
+      complete()
+      PushResult.Complete
+    } else {
+      super.push(data)
+    }
+  }
+
 }
 
-case class StreamingHttpResponseHeaderStub(version : HttpVersion, code : HttpCode, headers : Seq[(String, String)]) extends HttpResponseHeader
\ No newline at end of file
+/** A pipe designed to accept a chunked http body.
+ *
+ * The pipe will scan the chunk headers and auto-close itself when it hits the
+ * end of the stream (reading a 0 length chunk).  It also strips all chunk
+ * headers from the data, so the output stream is the raw data
+ *  
+ */
+class ChunkDecodingPipe extends InfinitePipe[DataBuffer] {
```

Comment:
Silly request:  can you move this near the ChunkEncodingPipe?  Just feels right :)